packages/neuron-ui/src/widgets/DropdownWithCustomRender/index.tsx

```diff
@@ -0,0 +1,191 @@
+    const newState = {
+      value,
+      label,
+      index,
+    }
+    fireChangeEvent(newState)
+    setSelectedValue(newState)
+    setIsOpen(false)
+  }
+
+  const handleDocumentClick = (event: MouseEvent | TouchEvent) => {
+    if (!dropdownRef.current?.contains(event.target as HTMLElement)) {
+      setIsOpen(false)
+    }
+  }
+
+  const handleMouseDown = (event: React.MouseEvent | React.TouchEvent) => {
+    if (typeof onFocus === 'function') {
+      onFocus(isOpen)
+    }
+    if (event.type === 'mousedown' && (event as React.MouseEvent<HTMLDivElement>).button !== 0) {
+      return
+    }
+    event.stopPropagation()
+    event.preventDefault()
+
+    if (!disabled) {
+      setIsOpen(openState => !openState)
+    }
+  }
+
+  const isValueSelected = selectedValue.value !== ''
+  const placeHolderValue = selectedValue.label || DEFAULT_PLACEHOLDER_STRING
+
+  const disabledClass = disabled ? `${styles['dropdown-disabled']}` : ''
+  const dropdownClass = `${styles['dropdown-root']} ${className} ${isOpen ? styles['is-open'] : ''}`
+  const controlClass = `${styles['dropdown-control']} ${disabledClass}`
+  const placeholderClass = `${styles['dropdown-placeholder']} ${isValueSelected ? styles['is-selected'] : ''}`
+  const arrowClass = `${styles['dropdown-arrow']}`
+  const menuClass = `${styles['dropdown-menu']}`
+
+  const renderOption = (option: OptionProps) => {
+    const { value, label, data, className: optionClassName, index } = option
+    const isSelected = value === selectedValue.value
+    const optionClass = `${styles['dropdown-option']} ${optionClassName} ${isSelected ? styles['is-selected'] : ''}`
+
+    const dataAttributes =
+      (data &&
+        Object.keys(data).reduce(
+          (acc, dataKey) => ({
+            ...acc,
+            [`data-${dataKey}`]: data[dataKey],
+          }),
+          {}
+        )) ||
+      {}
+    const renderedValue = {
+      value,
+      label,
+      index,
+    }
+
+    return (
+      <div
+        key={value}
+        className={optionClass}
+        onMouseDown={() => setValue(renderedValue)}
+        onClick={() => setValue(renderedValue)}
+        onKeyDown={() => setValue(renderedValue)}
+        role="option"
+        tabIndex={0}
+        aria-selected={isSelected ? 'true' : 'false'}
+        {...dataAttributes}
+      >
+        {label}
+      </div>
+    )
+  }
+
+  const buildMenu = () => {
+    const ops = optionsFromProps.map((option, index) => renderOption({ ...option, index }))
+
+    return ops.length ? ops : <div className={`${styles['dropdown-noresults']}`}>No options found</div>
+  }
+  const memoedMenu = useMemo(() => buildMenu(), [optionsFromProps])
+  const inputValue = <div className={placeholderClass}>{placeHolderValue}</div>
+  const menu = useMemo(
+    () =>
+      isOpen ? (
+        <div className={menuClass} aria-expanded="true">
+          {memoedMenu}
+        </div>
+      ) : null,
+    [isOpen, memoedMenu]
```

Comment:
Maybe you can compose the menu-created process into a function? It will be clear.